syntax = "proto3";

package product;
import "common.proto";

service ProductService {
  rpc GetProduct(ProductRequest) returns (ProductResponse);
  rpc ListProducts(ProductListRequest) returns (ProductListResponse);
  rpc CreateProduct(CreateProductRequest) returns (ProductResponse);
  rpc UpdateProduct(UpdateProductRequest) returns (ProductResponse);
}

message ProductRequest {
  string product_id = 1;
}

message ProductListRequest {
  common.Pagination pagination = 1;
  string category = 2;
}

message CreateProductRequest {
  string name = 1;
  string description = 2;
  double price = 3;
  int32 stock = 4;
  string category = 5;
  string image_url = 6;
}

message UpdateProductRequest {
  string product_id = 1;
  string name = 2;
  string description = 3;
  double price = 4;
  int32 stock = 5;
  string category = 6;
  string image_url = 7;
}

message ProductResponse {
  string product_id = 1;
  string name = 2;
  string description = 3;
  double price = 4;
  int32 stock = 5;
  string category = 6;
  string image_url = 7;
  string created_at = 8;
  string updated_at = 9;
}

message ProductListResponse {
  repeated ProductResponse products = 1;
  int32 total = 2;
  int32 page = 3;
  int32 limit = 4;
}
